### PR TITLE
Add Allow-/Block-List for Migrate & Mirrors

### LINKED ISSUE #4462: Repository migration Domains/IPs whitelist and blacklist
I would like to add options on app.ini to configure repository migration Domains/IPs whitelist and blacklist.

### PR DESCRIPTION
**Pull request by 6543**
This PR will add two options `blocklist` and `allowlist` for migrating repositories from external URLs.

When you set `allowlist` then `blocklist` will be ignored and all users on this gitea instance could only migrate repositories from the `domains` on `allowlist`.

When `allowlist` is empty and you have domains on `blocklist`, all users on this gitea instance could migrate repositories any domain except the `domains` on `blocklist`.

original pull: #8040
close #4462

TODOs:
 * [x] add default blocklist for private networks & localhost
 * [x] test manually to ensure it works

### CODE REVIEW

File: custom/conf/app.example.ini [hunk 1]
```diff
@@ -1212,3 +1212,11 @@ STORAGE_TYPE = local
 ;MINIO_LOCATION = us-east-1
 ; Minio enabled ssl only available when STORAGE_TYPE is `minio`
 ;MINIO_USE_SSL = false
+
+[migration]
+; Whitelist for migrating, default is blank. Blank means everything will be allowed.
```
Comment (6543) [inline custom/conf/app.example.ini]:
```suggestion
; Allowlist for migrating, default is blank. Blank means everything will be allowed.
```

File: custom/conf/app.example.ini [hunk 2]
```diff
@@ -1212,3 +1212,11 @@ STORAGE_TYPE = local
...
+; Multiple domains could be separated by commas.
+ALLOWLISTED_DOMAINS =
+; Blacklist for migrating, default is blank. Multiple domains could be separated by commas.
```
Comment (6543) [inline custom/conf/app.example.ini]:
```suggestion
; Blocklist for migrating, default is blank. Multiple domains could be separated by commas.
```

File: custom/conf/app.example.ini [hunk 3]
```diff
@@ -1212,3 +1212,11 @@ STORAGE_TYPE = local
...
+; When WHITELISTED_DOMAINS is not blank, this option will be ignored.
```
Comment (6543) [inline custom/conf/app.example.ini]:
```suggestion
; When ALLOWLISTED_DOMAINS is not blank, this option will be ignored.
```

File: docs/content/doc/advanced/config-cheat-sheet.en-us.md [hunk 4]
```diff
@@ -902,6 +902,11 @@ MINIO_USE_SSL = false
 
 And used by `[attachment]`, `[lfs]` and etc. as `STORAGE_TYPE`.
 
+## Migraions (`migration`)
+
+- `ALLOWLISTED_DOMAINS`: ****: Domains whitelist for migrating repositories, default is blank. It means everything will be allowed. Multiple domains could be separated by commas.
```
Comment (6543) [inline docs/content/doc/advanced/config-cheat-sheet.en-us.md]:
```suggestion
- `ALLOWLISTED_DOMAINS`: ****: Domains allowlist for migrating repositories, default is blank. It means everything will be allowed. Multiple domains could be separated by commas.
```

File: docs/content/doc/advanced/config-cheat-sheet.en-us.md [hunk 5]
```diff
@@ -902,6 +902,11 @@ MINIO_USE_SSL = false
...
+- `BLOCKLISTED_DOMAINS`: ****: Domains blacklist for migrating repositories, default is blank. Multiple domains could be separated by commas. When `ALLOWLISTED_DOMAINS` is not blank, this option will be ignored.
```
Comment (6543) [inline docs/content/doc/advanced/config-cheat-sheet.en-us.md]:
```suggestion
- `BLOCKLISTED_DOMAINS`: ****: Domains blocklist for migrating repositories, default is blank. Multiple domains could be separated by commas. When `ALLOWLISTED_DOMAINS` is not blank, this option will be ignored.
```

File: modules/matchlist/matchlist.go [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+package matchlist
+
+import (
+	"strings"
+
+	"github.com/gobwas/glob"
+)
+
+// Matchlist represents a black or white list
```
Comment (6543) [inline modules/matchlist/matchlist.go]:
```suggestion
// Matchlist represents a block or allow list
```

File: modules/matchlist/matchlist.go [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+type Matchlist struct {
+	rules     []string
+	ruleGlobs []glob.Glob
+}
+
+// NewMatchlist creates a new black or white list
```
Comment (6543) [inline modules/matchlist/matchlist.go]:
```suggestion
// NewMatchlist creates a new block or allow list
```

File: modules/setting/migrate.go [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+var Migration = struct {
+	AllowlistedDomains []string
+	BlocklistedDomains []string
+}{
+	AllowlistedDomains: []string{},
+	BlocklistedDomains: []string{},
+}
+
+// InitMigrationConfig represents load migration configurations
+func InitMigrationConfig() error {
```
Comment (6543) [inline modules/setting/migrate.go]: not used -> :eyes:

File: custom/conf/app.example.ini [hunk 9]
```diff
@@ -1187,6 +1187,14 @@ QUEUE_CONN_STR = "addrs=127.0.0.1:6379 db=0"
 MAX_ATTEMPTS = 3
 ; Backoff time per http/https request retry (seconds)
 RETRY_BACKOFF = 3
+; Allowlist for migrating, default is blank. Blank means everything will be allowed.
+; Multiple domains could be separated by commas.
+ALLOWLISTED_DOMAINS =
+; Blocklist for migrating, default is blank. Multiple domains could be separated by commas.
+; When ALLOWLISTED_DOMAINS is not blank, this option will be ignored.
+BLOCKLISTED_DOMAINS =
+; Allow private addresses defined by RFC 1918, RFC 1122, RFC 4632 and RFC 4291
```
Comment (6543) [inline custom/conf/app.example.ini]:
```suggestion
; Allow private addresses defined by RFC 1918, RFC 1122, RFC 4632 and RFC 4291 (false by default)
```

**Comment by 6543** [comment]
@lunny it's ready to look at :tada:

File: docs/content/doc/advanced/config-cheat-sheet.zh-cn.md [hunk 10]
```diff
@@ -313,6 +313,9 @@ IS_INPUT_FILE = false
 
 - `MAX_ATTEMPTS`: **3**: 在迁移过程中的 http/https 请求重试次数。
 - `RETRY_BACKOFF`: **3**: 等待下一次重试的时间，单位秒。
+- `ALLOWLISTED_DOMAINS`: **\<empty\>**: 迁移仓库的域名白名单，默认为空，表示允许从任意域名迁移仓库，多个域名用逗号分隔。
+- `BLOCKLISTED_DOMAINS`: **\<empty\>**: 迁移仓库的域名黑名单，默认为空，多个域名用逗号分隔。如果 `ALLOWLISTED_DOMAINS` 不为空，此选项将会被忽略。
+- `ALLOW_LOCALNETWORKS`: **false**: Allow private addresses defined by RFC 1918
```
Comment (6543) [inline docs/content/doc/advanced/config-cheat-sheet.zh-cn.md:318]: @a1012112796 if you like - do you have a suggestion?

File: custom/conf/app.example.ini [hunk 11]
```diff
@@ -1187,6 +1187,14 @@ QUEUE_CONN_STR = "addrs=127.0.0.1:6379 db=0"
...
 ; Backoff time per http/https request retry (seconds)
 RETRY_BACKOFF = 3
+; Allowlist for migrating, default is blank. Blank means everything will be allowed.
+; Multiple domains could be separated by commas.
+ALLOWLISTED_DOMAINS =
+; Blocklist for migrating, default is blank. Multiple domains could be separated by commas.
+; When ALLOWLISTED_DOMAINS is not blank, this option will be ignored.
+BLOCKLISTED_DOMAINS =
+; Allow private addresses defined by RFC 1918, RFC 1122, RFC 4632 and RFC 4291 (false by default)
+ALLOW_LOCALNETWORKS = false
```
Comment (silverwind) [inline custom/conf/app.example.ini:1197]: Hmm, these pref names seem a bit too generally named, how about `MIGRATE_ALLOWED_DOMAINS`, `MIGRATE_BLOCKED_DOMAINS` and `MIGRATE_ALLOWED_NETWORKS` (defaulting to a list of CIDRs), possibly also `MIGRATE_BLOCKED_NETWORKS`.

Comment (6543) [reply custom/conf/app.example.ini:1197]: @silverwind they are already in the `[migrations]` section so this should be fine ...

Comment (silverwind) [reply custom/conf/app.example.ini:1197]: hmm, I'd at least rename to `{BLOCKED,ALLOWED}_{DOMAINS,NETWORKS}`. I do think allowing to specify CIDRs would be quite nice to have too and domains could accept wildcards. Such parsing may be nice to have for other options too.

**Review by zeripath — APPROVED** [review]

**Comment by silverwind** [comment]
Still need to update `app.example.ini` with new names.

File: custom/conf/app.example.ini [hunk 12]
```diff
@@ -1187,6 +1187,14 @@ QUEUE_CONN_STR = "addrs=127.0.0.1:6379 db=0"
 MAX_ATTEMPTS = 3
 ; Backoff time per http/https request retry (seconds)
 RETRY_BACKOFF = 3
+; Allowlist for migrating, default is blank. Blank means everything will be allowed.
+; Multiple domains could be separated by commas.
+ALLOWLISTED_DOMAINS =
```
Comment (zeripath) [inline custom/conf/app.example.ini]:
```suggestion
ALLOWED_DOMAINS =
```

File: custom/conf/app.example.ini [hunk 13]
```diff
@@ -1187,6 +1187,14 @@ QUEUE_CONN_STR = "addrs=127.0.0.1:6379 db=0"
...
+; Blocklist for migrating, default is blank. Multiple domains could be separated by commas.
+; When ALLOWLISTED_DOMAINS is not blank, this option will be ignored.
+BLOCKLISTED_DOMAINS =
```
Comment (zeripath) [inline custom/conf/app.example.ini]:
```suggestion
BLOCKED_DOMAINS =
```

File: custom/conf/app.example.ini [hunk 14]
```diff
@@ -1187,6 +1187,14 @@ QUEUE_CONN_STR = "addrs=127.0.0.1:6379 db=0"
 MAX_ATTEMPTS = 3
 ; Backoff time per http/https request retry (seconds)
 RETRY_BACKOFF = 3
+; Allowlist for migrating, default is blank. Blank means everything will be allowed.
+; Multiple domains could be separated by commas.
+ALLOWLISTED_DOMAINS =
+; Blocklist for migrating, default is blank. Multiple domains could be separated by commas.
+; When ALLOWLISTED_DOMAINS is not blank, this option will be ignored.
```
Comment (zeripath) [inline custom/conf/app.example.ini]:
```suggestion
; When ALLOWED_DOMAINS is not blank, this option will be ignored.
```

File: custom/conf/app.example.ini [hunk 15]
```diff
@@ -1187,6 +1187,14 @@ QUEUE_CONN_STR = "addrs=127.0.0.1:6379 db=0"
 MAX_ATTEMPTS = 3
 ; Backoff time per http/https request retry (seconds)
 RETRY_BACKOFF = 3
+; Allowlist for migrating, default is blank. Blank means everything will be allowed.
```
Comment (zeripath) [inline custom/conf/app.example.ini]:
```suggestion
; Allowed domains for migrating, default is blank. Blank means everything will be allowed.
```

**Comment by zeripath** [comment]
> Still need to update `app.example.ini` with new names.

Is that now fixed?

**Comment by silverwind** [comment]
Yes those look fine now.

**Review by lunny — APPROVED** [review]

**Comment by 6543** [comment]
:rocket:

**Pull request merged**